text/3803-enum-repr-no-niches.md

```diff
@@ -0,0 +1,709 @@
+    SNOWY => println!("Do you wanna build a snowman?"),
+    weather => eprintln!("Unable to determine the weather, got enum value {}.", weather as u8),
+}
+```
+
+And would continue to work even if the `Weather::Snowy` variant was added in the future.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This adds surface area to the language for something that could be considered a relic from the C past.
+
+It may be confusing when to use this over `#[non_exhaustive]` (see also the syntax discussion in [the unresolved questions section][unresolved-questions]).
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Do nothing
+
+We could do nothing, and continue to recommend the `struct` + `const`s approach.
+
+## Add struct attribute
+
+We could add a helper attribute `#[diagnostics::enum_like]` on newtype `struct`s (or maybe on the `impl`) that could allow providing a user-experience more similar to what you get from regular enums.
+
+## Wait for pattern types, and use that in public API instead
+
+A common suggestion is to wait for [pattern types](https://internals.rust-lang.org/t/thoughts-on-pattern-types-and-subtyping/17675), define their interaction with niches, and instead suggest people write:
+
+```rust
+#[repr(u8)]
+pub enum Weather {
+    Sunny = 0,
+    Windy = 1,
+    Rainy = 2,
+    Unknown(u8 in 3..=255),
+}
+```
+
+Usage could then be:
+
+```rust
+match Weather::current() {
+    Weather::Sunny => println!("Nice an sunny!"),
+    Weather::Windy => println!("A bit breezy."),
+    Weather::Rainy => println!("Let's go inside and grab a cup of tea."),
+    Weather::Unknown(value) => eprintln!("Unable to determine the weather, got enum value {value}"),
+}
+```
+
+However, this is problematic, since the unknown variant can be matched on explicitly. Consider e.g.:
+
+```rust
+const SNOWY: Weather = Weather::Unknown(4);
+
+if let SNOWY = Weather::current() {
+    println!("Do you wanna build a snowman?");
+}
+```
+
+If the author of the Rust binding were to add `Weather::Snowy`, that would be a breaking change, regardless of the presence of `#[non_exhaustive]` (which makes it hard to use in bindings). The `Unknown` variant would have to be `#[doc(hidden)]` for that to work (and then, how would you then get the actual value out?), but avoiding such hacks was the motivation for adding `#[non_exhaustive]` in the first place.
+
+This is also a lot more error-prone, as you'd have to ensure that the niches in the pattern type does not overlap with the other variants. Consider e.g. the following innocuous change:
+
+```diff
+ #[repr(u8)]
+ pub enum Weather {
+     Sunny = 0,
+     Windy = 1,
+     Rainy = 2,
++    Snowy = 3,
+     Unknown(u8 in 3..=255),
+ }
+```
+
+This would be a fairly large bug, since the `Weather` enum suddenly grew to 2 bytes in size (instead, the `Unknown` variant should also have been updated to `u8 in 4..=255`).
```

Comment:
(Note that, as declared, `Weather` would already have been 2 bytes, since `#[repr(u8)]` explicitly separates the discriminant from the variants' fields, even if the discriminant could fit in a field's niche)

Reply:
Good point! I've changed the text as you suggested.